File: algorithms/Python/sorting/radix_sort.py
"""
Radix Sort
Time Complexity: O(nk + n). 
n is the size of input list and k is the digit length of the number.
"""

def radix_sort(arr, simulation=False):
    position = 1
    max_number = max(arr)

    iteration = 0
    if simulation:
        print("iteration", iteration, ":", *arr)

    while position <= max_number:
        queue_list = [list() for _ in range(10)]

        for num in arr:
            digit_number = num // position % 10
            queue_list[digit_number].append(num)

        index = 0
        for numbers in queue_list:
            for num in numbers:
                arr[index] = num
                index += 1

        if simulation:
            iteration = iteration + 1
            print("iteration", iteration, ":", *arr)

        position *= 10
    return arr

arr = [142, 34, 6, 17, 113, 2, 261, 72]

if __name__ == "__main__":
    print("Sorted Array:", radix_sort(arr))
    
# Runtime Test Cases:-
# Test case 1.
# Enter the list of numbers: 12 7 56 42
# Sorted list: [7 12 42 56]
# Test case 2.
# Enter the list of numbers: 142 34 6 17 113 2 261 72 
# Sorted list: [2 6 17 34 72 113 142 261]
# For radix sort, Best = Worst = Average Time Complexity is O(nk+n).

File: algorithms/C/sorting/radix-sort.c
#include <stdio.h>


int get_max( int arr[], int n){

    int myMax = -1 ;
    for(int i=0; i<n; i++){
        if(arr[i] > myMax)myMax = arr[i] ;
    }

    return myMax ;
}

void count_sort(int arr[], int n, int digit){
    int count[10] = {0} ;
    int *sorted ;
    sorted = (int*)malloc(n * sizeof(int));

    for(int i=0; i<n; i++){
        count[(arr[i]/digit)%10]++ ;
    }
    for(int i=1; i<10; i++){
        count[i] += count[i-1] ;
    }

    for(int i=n-1; i>=0; i--){
        sorted[--count[(arr[i]/digit)%10]] = arr[i] ;
    }

    for(int i=0; i<n; i++){
        arr[i] = sorted[i] ;
    }
}

void radix_sort(int arr[], int n){

    int myMax = get_max(arr,n) ;

    for(int digit=1; myMax/digit; digit*=10){
        count_sort(arr,n,digit) ;
    }
}

int main() {
    //                  :  The best :  The avg : The worst  : Space 
    //       complexity :    O(kn)  :   O(kn)  :  O(kn)     :  O(d+n)

	//         Stability :   Inplace    :      Comparison Sort 
        //          Yes      :     Yes      :       No 

    int arr[] = {374, 42, 6, 920,825} ;
    int n = sizeof(arr)/sizeof(int) ;

    printf("Before sorting : ") ;
    for(int i=0; i<n; i++){
        printf("%d ",arr[i] );
    }

    radix_sort(arr, n) ;

    printf("\n\nAfter sorting : ") ;
    for(int i=0; i<n; i++){
        printf("%d ",arr[i] );
    }
    return 0;
}
